src/ert/storage/local_ensemble.py

```diff
@@ -885,73 +909,118 @@ def get_observations_and_responses(
-
-            observations_for_type = observations_by_type[response_type].filter(
-                polars.col("observation_key").is_in(list(selected_observations))
-            )
-            responses_for_type = self.load_responses(
-                response_type, realizations=tuple(iens_active_index)
-            )
-
-            # Note that if there are duplicate entries for one
-            # response at one index, they are aggregated together
-            # with "mean" by default
-            pivoted = responses_for_type.pivot(
-                on="realization",
-                index=["response_key", *response_cls.primary_key],
-                aggregate_function="mean",
-            )
-
-            # We need to either assume that if there is a time column
-            # we will approx-join that, or we could specify in response configs
-            # that there is a column that requires an approx "asof" join.
-            # Suggest we simplify and assume that there is always only
-            # one "time" column, which we will reindex towards the response dataset
-            # with a given resolution
-            if "time" in pivoted:
-                joined = observations_for_type.join_asof(
-                    pivoted,
-                    by=["response_key", *response_cls.primary_key],
-                    on="time",
-                    tolerance="1s",
-                )
-            else:
-                joined = observations_for_type.join(
-                    pivoted,
-                    how="left",
-                    on=["response_key", *response_cls.primary_key],
+        with polars.StringCache():
+            dfs_per_response_type = []
+            for (
+                response_type,
+                response_cls,
+            ) in self.experiment.response_configuration.items():
+                if response_type not in observations_by_type:
+                    continue
+
+                observations_for_type = (
+                    observations_by_type[response_type]
+                    .filter(
+                        polars.col("observation_key").is_in(list(selected_observations))
+                    )
+                    .with_columns(
+                        [
+                            polars.col("response_key")
+                            .cast(polars.Categorical)
+                            .alias("response_key")
+                        ]
+                    )
                 )
 
-            joined = (
-                joined.with_columns(
-                    polars.concat_str(response_cls.primary_key, separator=", ").alias(
-                        "__tmp_index_key__"  # Avoid potential collisions w/ primary key
+                observed_cols = {
+                    k: observations_for_type[k].unique()
+                    for k in ["response_key", *response_cls.primary_key]
+                }
+
+                reals = iens_active_index.tolist()
+                reals.sort()
```

Comment:
Do we need explicit sort?

Reply:
Not 100%, sure, but if we do a sort it costs us nothing, and it ensures that we append the realization-columns in ascending order. If we are given a non-sorted `iens_active_index` there are no checks that would catch it afaik. Though all `test_es_update.py` cli / unit tests do still pass without the sort.